### PR TITLE
[SYCL][Doc] Add device global extension spec

### PR DESCRIPTION
**Pull request by mkinsner**
Add device global extension spec

Signed-off-by: Michael Kinsner <michael.kinsner@intel.com>

### CODE REVIEW

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (Pennycook) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:232]: Nit: Using upper camel case (i.e. `PropertyListT`) for template arguments would help avoid confusion with existing typenames, and makes it more obvious when something else in the class definition is referring back to a template parameter.  I think it's especially important here, because the similarly named `property_list_t` exists in the `SYCL_EXT_ONEAPI_PROPERTY_LIST` extension.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (Pennycook) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:284]: Nit: Same comment as above: I recommend renaming `propertyT` to `PropertyT`.  There are other instances of `propertyT` in the document, but I won't highlight them all.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (Pennycook) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:432]: Reading through the table beneath I was still surprised by the presence of the `startIndex` and `offset` arguments.  Maybe a non-normative `NOTE:` here that reminds readers that there's no way to perform pointer arithmetic directly on the `device_global` would be useful here?

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (Pennycook) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:766]:
```suggestion
*Resolved*: Yes, but only on the device side.  There is no visibility/communication across devices because each device receives a unique allocation of type _T_ underlying the `device_global`.  There is no way for an `atomic_ref` associated with the allocation to be created in host code because there is no way to extract a pointer or reference in host code (only copy/memcpy).
```

**Review by Pennycook — COMMENTED** [review]
A few minor formatting suggestions, but on the whole LGTM.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (mkinsner) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc]: Agreed.  Fixed.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (mkinsner) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:284]: Fixed.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (mkinsner) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc]: Fixed.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (mkinsner) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:432]: Agreed.  Added a note.

**Review by Pennycook — APPROVED** [review]

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (artemrad) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:208]: This is already covered above. No new details are introduced. Should be removed.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (artemrad) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:766]: "on the same device", is more explicit and clearer

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (keryell) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:201]:
```suggestion
    h.single_task([=] {
```

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (keryell) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:753]:
```suggestion
    h.single_task([=] {
```

**Review by keryell — COMMENTED** [review]
Quite useful and simple to use!
This is simpler than https://github.com/triSYCL/triSYCL/blob/master/tests/scope/queue.cpp proposal and the fact you can access globally a variable is very powerful to avoid passing handler or context all over the place.
Do you have also a pure host library implementation not requiring any specific compiler?

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (mkinsner) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc]: Fixed

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (mkinsner) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc]: Fixed

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (mkinsner) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:136]: `operator->` is overloaded such that if the `device_global` contains a pointer to a type where `operator->` is valid, then `->` will work as expected.  `device_global` has similar semantics to a reference wrapper, and since the dot operator can't be overloaded, we end up needing `.get()` if the `device_global` contains a struct directly (not pointer to it).

Comment (keryell) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:136]: Great! That makes sense.
@erichkeane @AaronBallman do you remember where the dot operator got distracted in ISO C++? Just the overflow of more urgent proposals?

Comment (AaronBallman) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:136]: @keryell -- It didn't have consensus to proceed under the form proposed. I wouldn't be surprised if there were additional proposals in the future, but AFAIK, nothing is planned.

Comment (keryell) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:136]: @AaronBallman thanks for the feedback. Too bad it got stuck.

**Comment by mkinsner** [comment]
> Quite useful and simple to use!
> This is simpler than https://github.com/triSYCL/triSYCL/blob/master/tests/scope/queue.cpp proposal and the fact you can access globally a variable is very powerful to avoid passing handler or context all over the place.
> Do you have also a pure host library implementation not requiring any specific compiler?

Our original forms of the extension were designed to be straightforward to implement with a pure library, but the resultant syntax was making the feature too cumbersome to use (e.g. passing device handler through call trees) and broke the value proposition in multiple cases.  We decided to prioritize usability with this version, and removed the language aspects that aided lib-only implementation.  It's likely still possible to implement that way but with additional complexity - we concluded that clean language was more important in this feature based on feedback and by looking at the motivating uses.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:297]:
```suggestion
multi_ptr<T, access::address_space::global_space, IsDecorated>
  get_multi_ptr() noexcept;
```

The HTML render for this table causes the contents of the right column to be cut off.  I suspect this is because the left column it too wide.  I suspect it may help to split these long lines as I show above.  Same for the long declaration below.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:377]:
```suggestion
| Returns true if the `PropertyListT` contains the property specified by `propertyT`. Returns false if it does not.
```

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:388]:
```suggestion
Must produce a compiler diagnostic if `PropertyListT` does not contain a `propertyT` property.
```

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:438]:
```suggestion
    size_t count, size_t startIndex,
    const std::vector<event> &depEvents);
```

**Comment by keryell** [comment]
> Our original forms of the extension were designed to be straightforward to implement with a pure library, but the resultant syntax was making the feature too cumbersome to use (e.g. passing device handler through call trees) and broke the value proposition in multiple cases. We decided to prioritize usability with this version, and removed the language aspects that aided lib-only implementation. It's likely still possible to implement that way but with additional complexity - we concluded that clean language was more important in this feature based on feedback and by looking at the motivating uses.

That makes sense.
Probably playing with TLS would allow a pure library implementation.

**Review by jbrodman — DISMISSED** [review]

**Comment by bader** [comment]
It looks like there a few comments from @gmlueck to address. @mkinsner, @gmlueck, am I right?

**Comment by gmlueck** [comment]
> It looks like there a few comments from @gmlueck to address. @mkinsner, @gmlueck, am I right?

Yes, and we agreed recently on a few more changes too.  @mkinsner asked me to make the changes myself, so I will create a PR against this PR, and mike will merge my PR into his.  After that, we can merge mike's PR into the mainline.

**Comment by bader** [comment]
Thanks for the update.
Just in case it might be useful, I think you should be able to commit directly to https://github.com/mkinsner/llvm/tree/private/mkinsner/device_global_extension.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (gmlueck) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc:54]: We decided that we will remove this language with the commit that adds the implementation.

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (gmlueck) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc]: Done in 3e04b444abe7b8600880790d41c976cdad7d351c

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (gmlueck) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc]: Done in 3e04b444abe7b8600880790d41c976cdad7d351c

File: sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc
Comment (gmlueck) [reply sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc]: Done in 3e04b444abe7b8600880790d41c976cdad7d351c

**Comment by gmlueck** [comment]
> It looks like there a few comments from @gmlueck to address. @mkinsner, @gmlueck, am I right?

@mkinsner told me verbally that he agrees with my comments and asked me to make them myself.  I did this in 3e04b444abe7b8600880790d41c976cdad7d351c.

I approve this PR now, and it can be merged unless others have concerns.

**Review by gmlueck — DISMISSED** [review]

**Review by gmlueck — APPROVED** [review]

**Review by jbrodman — APPROVED** [review]

**Comment by gmlueck** [comment]
@olegmaslovatintel: Can someone merge this PR while @bader is on vacation?  I think all outstanding issues have been resolved.

**Comment by olegmaslovatintel** [comment]
@vladimirlaz  @pvchupin  could you please merge?

**Pull request merged**
